### PR TITLE
Partial roll-forward of "[Typography] Migrate from FontScheme to TypographyScheming (#3219)"

### PR DESCRIPTION
**Pull request by jverkoey**
This change only adds the new Typography Scheme APIs and target with no component modifications.

This partially reverts commit 83b9ea1735a6d38359896e79452713bd85341ef3.

Pivotal story: https://www.pivotaltracker.com/story/show/156392054 and https://www.pivotaltracker.com/story/show/156392171

### CODE REVIEW

File: components/schemes/Typography/src/MDCTypographyScheme.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+#import <UIKit/UIKit.h>
+
+/**
+ A simple typography scheme that provides semantic fonts. There are no optional
+ properties, all fonts must be provided, supporting more reliable typography theming.
+ */
+@protocol MDCTypographyScheming
+
+@property(nonatomic, nonnull, readonly) UIFont *headline1;
```
Comment (romoore) [inline components/schemes/Typography/src/MDCTypographyScheme.h:25]: Please provide documentation for these properties.

File: components/schemes/Typography/src/MDCTypographyScheme.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+@property(nonatomic, nonnull, readwrite) UIFont *subtitle1;
+@property(nonatomic, nonnull, readwrite) UIFont *subtitle2;
+@property(nonatomic, nonnull, readwrite) UIFont *body1;
+@property(nonatomic, nonnull, readwrite) UIFont *body2;
+@property(nonatomic, nonnull, readwrite) UIFont *caption;
+@property(nonatomic, nonnull, readwrite) UIFont *button;
+@property(nonatomic, nonnull, readwrite) UIFont *overline;
+
+/**
+ Convenience initializer that calls @c initWithMaterialDefaults.
```
Comment (romoore) [inline components/schemes/Typography/src/MDCTypographyScheme.h]: Based on your PR (#3246), prefer different docs: "Initializes the typography scheme with the latest material defaults."

File: components/schemes/Typography/src/MDCTypographyScheme.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ */
+- (nonnull instancetype)init;
+
+/**
+ Initializes an instance of MDCSemanticColorScheme with the Material default fonts.
```
Comment (romoore) [inline components/schemes/Typography/src/MDCTypographyScheme.h]: Based on your PR (#3246), prefer different docs: "Initializes the typography scheme with the Material defaults, circa April 2018."

Also, accidental copy of `MDCSemanticColorScheme`.

File: components/schemes/Typography/src/MaterialTypographyScheme.h [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*
- Copyright 2017-present the Material Components for iOS authors. All Rights Reserved.
+ Copyright 2018-present the Material Components for iOS authors. All Rights Reserved.
```
Comment (romoore) [inline components/schemes/Typography/src/MaterialTypographyScheme.h:2]: Don't change copyrights when renaming?

File: components/schemes/Typography/tests/unit/MDCTypographySchemeTests.m [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  XCTAssertEqual(self.scheme.headline6, defaultScheme.headline6);
+  XCTAssertEqual(self.scheme.subtitle1, defaultScheme.subtitle1);
+  XCTAssertEqual(self.scheme.subtitle2, defaultScheme.subtitle2);
+  XCTAssertEqual(self.scheme.body1, defaultScheme.body1);
+  XCTAssertEqual(self.scheme.body2, defaultScheme.body2);
+  XCTAssertEqual(self.scheme.caption, defaultScheme.caption);
+  XCTAssertEqual(self.scheme.button, defaultScheme.button);
+  XCTAssertEqual(self.scheme.overline, defaultScheme.overline);
+}
+
```
Comment (romoore) [inline components/schemes/Typography/tests/unit/MDCTypographySchemeTests.m]: Please add a test to verify that `-init` and `-initWithMaterialDefaults` produce equivalent typography schemes.

**Review by romoore — DISMISSED** [review]

File: components/schemes/Typography/src/MDCTypographyScheme.h (hunk 1 above)
Comment (jverkoey) [reply components/schemes/Typography/src/MDCTypographyScheme.h:25]: @ajsecord is there documentation snippets for these APIs?

File: components/schemes/Typography/src/MDCTypographyScheme.h (hunk 2 above)
Comment (jverkoey) [reply components/schemes/Typography/src/MDCTypographyScheme.h]: Done.

File: components/schemes/Typography/src/MDCTypographyScheme.h (hunk 3 above)
Comment (jverkoey) [reply components/schemes/Typography/src/MDCTypographyScheme.h]: Done.

File: components/schemes/Typography/src/MaterialTypographyScheme.h (hunk 4 above)
Comment (jverkoey) [reply components/schemes/Typography/src/MaterialTypographyScheme.h:2]: This isn't a rename. git is mistaken.

File: components/schemes/Typography/tests/unit/MDCTypographySchemeTests.m (hunk 5 above)
Comment (jverkoey) [reply components/schemes/Typography/tests/unit/MDCTypographySchemeTests.m]: Done.

File: components/schemes/Typography/src/MDCTypographyScheme.h (hunk 1 above)
Comment (ajsecord) [reply components/schemes/Typography/src/MDCTypographyScheme.h:25]: Er, this is getting close to an emergency in terms of our release, so let's prioritize getting this in and add docs later. Add an issue and assign to the original author?

**Review by ajsecord — APPROVED** [review]

File: components/schemes/Typography/src/MDCTypographyScheme.h (hunk 1 above)
Comment (jverkoey) [reply components/schemes/Typography/src/MDCTypographyScheme.h:25]: +1

https://www.pivotaltracker.com/story/show/156555861

File: components/schemes/Typography/src/MDCTypographyScheme.m [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import "MDCTypographyScheme.h"
+
+#import <CoreText/CoreText.h>
```
Comment (romoore) [inline components/schemes/Typography/src/MDCTypographyScheme.m:19]: Nit: Is CoreText required? I don't see it in the BUILD file, so I'm guessing it's not necessary? Feel free to remove it in the follow-up PR for docs.

**Review by romoore — APPROVED** [review]

File: components/schemes/Typography/src/MDCTypographyScheme.m (hunk 6 above)
Comment (jverkoey) [reply components/schemes/Typography/src/MDCTypographyScheme.m:19]: filed: https://www.pivotaltracker.com/story/show/156561429

**Pull request merged**